src/coverlet.template/ModuleTrackerTemplate.cs

```diff
@@ -50,61 +56,76 @@ public static void UnloadModule(object sender, EventArgs e)
+                            {
+                                var intPointer = (int*) pointer;
 
-                        for (int i = 0; i < hitsLength; ++i)
-                        {
-                            int oldHitCount = br.ReadInt32();
-                            bw.Seek(-sizeof(int), SeekOrigin.Current);
-                            bw.Write(HitsArray[i] + oldHitCount);
+                                // Signal back to coverage analysis that we've started transferring hit counts.
+                                // Use interlocked here to ensure a memory barrier before the Coverage class reads
+                                // the shared data.
+                                Interlocked.Increment(ref *(intPointer + HitsResultUnloadStarted));
+
+                                for (var i = 0; i < HitsArray.Length; i++)
+                                {
+                                    var count = HitsArray[i];
+
+                                    if (count > 0)
+                                    {
+                                        var hitLocationArrayOffset = intPointer + i + HitsResultHeaderSize;
+
+                                        // No need to use Interlocked here since the mutex ensures only one thread updates 
+                                        // the shared memory map.
+                                        *hitLocationArrayOffset += count;
+                                    }
+                                }
+
+                                // Signal back to coverage analysis that all hit counts were successfully tallied.
+                                Interlocked.Increment(ref *(intPointer + HitsResultUnloadFinished));
+                            }
+                            finally
+                            {
+                                buffer.ReleasePointer();
+                            }
                         }
                     }
                 }
-
-                // Prevent any double counting scenario, i.e.: UnloadModule called twice (not sure if this can happen in practice ...)
-                // Only an issue if DomainUnload and ProcessExit can both happens, perhaps can be removed...
-                Array.Clear(HitsArray, 0, HitsArray.Length);
-
-                // On purpose this is not under a try-finally: it is better to have an exception if there was any error writing the hits file
-                // this case is relevant when instrumenting corelib since multiple processes can be running against the same instrumented dll.
-                mutex.ReleaseMutex();
+                finally
+                {
+                    // Avoid double counting if unloaded multiple times (if that can happen?)
+                    HitsArray = new int[HitsArray.Length];
+                    
+                    mutex.ReleaseMutex();
+                    memoryMap?.Dispose();
+                }
             }
+
+            s_isTracking = false;
```

Comment:
Here the no-longer threadstatic field is set to false, in the assumption that UnloadModule is not called in parallel in the same appdomain. This is necessary for the multiple-call test case, but if that assumption is not true (cf #291 which changed it from treadstatic) another solution could be found.

Reply:
Nevermind, @sharwell removes this field entirely in #297... So that should go in to master first, then that change needs to be merged into this PR before it is (again) ready to merge.

Reply:
@petli His change is already merged in master.